client/java/src/main/java/com/linecorp/centraldogma/client/armeria/CentralDogmaEndpointGroup.java

```diff
@@ -0,0 +1,141 @@
+ * {@link Endpoint}s from a route file served by CentralDogma, and update the list when upstream data changes.
+ * Route file could be json file or normal text file.
+ * <p>
+ * <p>In below example, json file with below content will be served as route file:
+ * <p>
+ * <pre>{@code
+ *  [
+ *      "host1:port1",
+ *      "host2:port2",
+ *      "host3:port3"
+ *  ]
+ * }
+ * </pre>
+ * <p>
+ * <p>The route file could be retrieve as {@link EndpointGroup} using below code:
+ * <p>
+ * <pre>{@code
+ *  CentralDogmaEndpointGroup<JsonNode> endpointGroup = CentralDogmaEndpointGroup.ofJsonFile(
+ *      centralDogma, "myProject", "myRepo",
+ *      CentralDogmaCodec.DEFAULT_JSON_CODEC,
+ *      Query.ofJsonPath("/route.json")
+ *  )
+ *  endpointGroup.endpoints();
+ * }
+ * </pre>
+ *
+ * @param <T> Type of CentralDomgma file (could be JsonNode or String)
+ */
+public final class CentralDogmaEndpointGroup<T> extends DynamicEndpointGroup {
+    private static final Logger logger = LoggerFactory.getLogger(CentralDogmaEndpointGroup.class);
+    private static final long WATCH_INITIALIZATION_TIMEOUT_MILLIS = TimeUnit.SECONDS.toMillis(10);
+
+    private final Watcher<T> instanceListWatcher;
+    private final EndpointListCodec<T> endpointCodec;
+
+    /**
+     * Creates a new {@link CentralDogmaEndpointGroup}.
+     *
+     * @param endpointCodec A {@link EndpointListCodec}
+     * @param watcher A {@link Watcher}
+     */
+    public static <T> CentralDogmaEndpointGroup<T> ofWatcher(EndpointListCodec<T> endpointCodec,
+                                                             Watcher<T> watcher) {
+        return new CentralDogmaEndpointGroup<>(watcher, endpointCodec);
+    }
+
+    /**
+     * Creates a new {@link CentralDogmaEndpointGroup}.
+     * @param uri an uri of CentralDogma server
+     * @param dogmaProject CentralDogma project name
+     * @param dogmaRepo CentralDogma repository name
+     * @param dogmaCodec A {@link CentralDogmaCodec}
+     * @param waitTime a {@code long} define how long we should wait for initial result
+     * @param waitUnit a {@link TimeUnit} define unit of wait time
+     * @throws CentralDogmaEndpointException if couldn't get initial result from CentralDogma server
+     */
+
+    /**
+     * Creates a new {@link CentralDogmaEndpointGroup}.
+     *
+     * @param centralDogma A {@link CentralDogma}
+     * @param endpointCodec A {@link EndpointListCodec}
+     * @param projectName CentralDogma project name
+     * @param repositoryName CentralDogma repository name
+     * @param query A {@link Query} to route file
+     */
+    public static <T> CentralDogmaEndpointGroup<T> of(CentralDogma centralDogma,
+                                                      EndpointListCodec<T> endpointCodec,
+                                                      String projectName, String repositoryName,
+                                                      Query<T> query) {
```

Comment:
I feel like it looks better moving `endpointCodec` to the end of the parameter list.